Skip session recording reservation files (filessesion)

gabrielcorado opened the pull request:
Closes #13640 by skipping reservation files when completing a session recording upload. To do so, now we create the reservation files with a different extension so we can differentiate them from regular upload parts.

### Context
[Session recording modes](https://github.com/gravitational/teleport/pull/12916) introduce ["disk reservation"](https://github.com/gravitational/teleport/blob/master/lib/events/api.go#L650-L652), where Teleport creates placeholder files to guarantee that the node has enough disk space to record sessions. This reservation happens every time a new part file is required. Those files are replaced with recording contents on the [`UploadPart`](https://github.com/gravitational/teleport/blob/master/lib/events/filesessions/filestream.go#L106) flow.

In scenarios where the upload couldn't run (for example, if the node is killed during a session), the reservation file was left on the node. When the [`UploadComplete`](https://github.com/gravitational/teleport/blob/master/lib/events/complete.go#L143) was executed, it created broken recording files (`.tar`), causing the uploads to fail (as we can see in the logs presented on the related issue).

gabrielcorado committed:
feat(filesessions): track reservation files

zmb3 commented on lib/events/filesessions/filestream.go:
```diff
@@ -384,6 +395,46 @@ func checkUploadID(uploadID string) error {
+
+	// Encode reservation content.
+	encoded := &bytes.Buffer{}
+	encoder := gob.NewEncoder(encoded)
```
What is going on here? Why do we need to introduce yet another encoding format?

gabrielcorado replied:
I wasn't sure what encoding to use to identify those "reservation" files. First, I've added a Protobuf message, but it seemed like a little of an overkill for this since we're not using the data inside those messages. Do you think we should still use Protobuf for this, or do you have any other suggestions? The idea is to have some content on those files to differentiate them from regular session recording parts (Protobuf encoded).

espadolini replied:
Why not just rename them to `<uuid>.reserve` or something along those lines?

gabrielcorado replied:
> Why not just rename them to <uuid>.reserve or something along those lines?

I think that it is a good alternative because the [`ListParts`](https://github.com/gravitational/teleport/blob/3ffe935f478e51a339257464e5de0f0e28e398b6/lib/events/filesessions/filestream.go#L353) wouldn't recognize those files (not causing it to create broken session recordings) and they will be removed by the [`CompleteUpload`](https://github.com/gravitational/teleport/blob/master/lib/events/filesessions/filestream.go#L184).

gabrielcorado committed:
chore(filesession): fix typo

espadolini commented:
Side note: don't we lose the disk space reservation the moment we open the file for writing again in  [`openUploadPart`](https://github.com/gravitational/teleport/blob/3ffe935f478e51a339257464e5de0f0e28e398b6/lib/events/filesessions/filestream.go#L321)?

espadolini commented:
> Do you mean it because we truncate the file when opening?

Yes, as opposed to writing whatever content we need and then truncating before closing.

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files

gabrielcorado committed:
refactor(filesessions): change reservation file extension

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files

gabrielcorado commented:
@zmb3 I've updated the PR with @espadolini suggestion: create the reservation files with a different extension, so we don't need to check their content.

> Yes, as opposed to writing whatever content we need and then truncating before closing.
I have updated the PR with this also. Thanks for pointing it out.

espadolini reviewed (approved):
Never mucked around the session recording code so take this approval with a grain of salt

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/remove-reservation-files
